### DISCUSSION TITLE
Moving the robot in the bed bath task

### DISCUSSION THREAD

**Discussion by rajas1310** [empriselab/RCareWorld#109] [Help - Using] [answered]
Hi, 
I was trying to rotate the robot base in the Bed Bath environment using the functions `TurnLeft(angle,speed)` and `TurnRight(angle,speed)`. 
When I input the angle as 180, the Stretch 3 turns only by  90 degrees.
When I input the angle as 90, the Stretch 3 turns approximately by  45 degrees. 
When I input the angle as 270, the Stretch 3 turns by more than 90 degrees and makes very jerky movements ultimately falling down.

The speed parameter did not make any difference as per my observation.

Is there something that I have missed in understanding what those (`angle` and `speed`) arguments do? Or is it a bug?

PFA the video for Case-1 (angle=180, speed=60) and Case-2 (angle=270, speed=60)  
[rotation-180-60.webm](https://github.com/user-attachments/assets/b319fc64-e05c-48be-b65e-a8486f6c9f72)
[rotation-270-60.webm](https://github.com/user-attachments/assets/58758c85-900a-445a-962e-8675e3def375)

**Comment by YoruCathy** [comment]
We've updated the scripts in the newest commit https://github.com/empriselab/RCareWorld/commit/0953f703e081cfbf23b919826032dc727085c0a9
Could you pull it and give it a try?

**Reply by rajas1310** [reply]
Hey, thank you for the update.

I just tried the updated script, but the problem still exists. The robot doesn't turn full 180 degrees when I input `TurnLeft(180,2.0)`.
But, I can observe a tiny difference as I change the speed argument now.

**Reply by YoruCathy** [reply]
Hi @rajas1310 this is updated in https://github.com/empriselab/RCareWorld/commit/48aad59cb990e828205c369a879f8966b1c819ec
We recommend avoiding moving the arm all the way to the upper extreme, which might cause the robot to fall. We also recommend making small, incremental movements and turns, followed by moving the gripper after completing these movements.

**Reply by rajas1310** [reply]
Hi, thanks for the suggestive comment in the `test_bathing.py` as well. 
But, Stretch is not moving when I run that same script. Particularly, the Move and Turn functions aren't working. I tried changing speed from 1 to 1000.

**Reply by shxingch** [reply]
Hi @rajas1310 ! When you attempt to turn at large angles, remember to adjust the orientation of the gripper and change it in time; otherwise, the default position of the gripper will conflict with the robot's rotation, causing the turn to fail. 
We recommend that you make small movements, change the orientation of the gripper after each move, and then repeat the process. This will allow for more stable motion of the robot!

**Reply by rajas1310** [reply]
Okay, but the robot is not moving forward or backward too.
And in case of making a turn, I am trying to turn only by 90 deg so I believe it should work.

The following snippet from `test_bathing.py` (file from the pulled branch) should ideally make the robot turn left by 90, then right by 90, and finally move forward. But the base of the robot hardly moves. 

```
    robot.TurnLeft(90, 1)
    env.step(600)
    
    # robot.StopMovement()
    # env.step(30)
    
    robot.TurnRight(90, 1)
    env.step(600)
    
    # robot.StopMovement()
    # env.step(30)
    
    robot.MoveForward(0.6, 0.2)
    env.step(300)

```
Could you please give another demo code where it will be clear on how to move the robot at large angles, forward and backward while adjusting the gripper?

**Reply by YoruCathy** [reply]
I went through the past versions, and[ this one](https://github.com/empriselab/RCareWorld/commit/c7f4a7cdf82ff0cb0812548d6c0a5e3aea9a398b) should be working. We had unit [tests](https://github.com/empriselab/RCareWorld/commit/c7f4a7cdf82ff0cb0812548d6c0a5e3aea9a398b#diff-f5f77ce5629f7153bf232c7990329388eabc28642bdadb30d2b8b34ce6e5a3a6) for Stretch movement. This commit is passing all the tests. I noticed the newest commits is directly committed instead of opening a pull request, so this [commit](https://github.com/empriselab/RCareWorld/commit/25c9eab30803ac610dc279adf8a5e0f3cfe3f0b8) didn't trigger the unit tests and broke the simulation. For now, you can probably use the commit that's working if you don't need to use getBBox for the sponge. We'll fix it as soon as possible.

**Reply by rajas1310** [reply]
Thank you!

**Reply by martinschonger** [reply]
For the unit tests, when I compare `new_robot_base_rotation` and `expected_robot_base_rotation` on line https://github.com/empriselab/RCareWorld/blob/6f0b816372c12449f8df2d4ec387722e283fb1b5/tests/test_bathing_env.py#L83, they don't match up at all (taking into account the +- 5 deg). Still, the test passes. I believe on line https://github.com/empriselab/RCareWorld/blob/6f0b816372c12449f8df2d4ec387722e283fb1b5/tests/test_bathing_env.py#L38 it should be `atol=degrees_to_radians(atol)`. Further, I believe it should be `(0.0, -90.0, 0.0)` on line https://github.com/empriselab/RCareWorld/blob/6f0b816372c12449f8df2d4ec387722e283fb1b5/tests/test_bathing_env.py#L80 (and similarly for `TurnRight`). These changes render the test as 'failed'.

I did some testing with the base rotation, and just can't get it to work properly. For example, the code
```python
import numpy as np

robot_base_rotation = stretch.data["rotations"][0].copy()
print(robot_base_rotation)
stretch.TurnRight(20, 1)
env.step(600)
new_robot_base_rotation = stretch.data["rotations"][0].copy()
print(new_robot_base_rotation)
expected_robot_base_rotation = np.add(robot_base_rotation, (0.0, 20.0, 0.0))
print(expected_robot_base_rotation)

robot_base_rotation = stretch.data["rotations"][0].copy()
print(robot_base_rotation)
stretch.TurnLeft(20, 1)
env.step(600)
new_robot_base_rotation = stretch.data["rotations"][0].copy()
print(new_robot_base_rotation)
expected_robot_base_rotation = np.add(robot_base_rotation, (0.0, -20.0, 0.0))
print(expected_robot_base_rotation)
```
produces the following output:
```
[357.990478515625, 359.9644470214844, 0.00018981154426001012]
[357.7393798828125, 16.252723693847656, 0.05774369463324547]
[3.57990479e+02 3.79964447e+02 1.89811544e-04]

[357.7393798828125, 16.252723693847656, 0.05774369463324547]
[357.976806640625, 15.646373748779297, 359.9814758300781]
[ 3.57739380e+02 -3.74727631e+00  5.77436946e-02]
```
Specifically, the left turn does not work at all here. Why is that?

**Reply by YoruCathy** [reply]
Hi @martinschonger, first thing, Unity's coordinate system is not the same as the usual coordinate system you see in common robotics software such as Gym, etc. For example, https://techarthub.com/a-guide-to-unitys-coordinate-system-with-practical-examples/ so you need to consider this difference. The Euler angle data should be in degrees, not radians. Due to the physics modeling (friction, etc), it's not that accurate for the moving and turning control. I'd suggest you try tuning the speed, distance move/degree to rotate, and other stuff, and then come up with a higher-level controller.

**Reply by martinschonger** [reply]
Thanks for your reply, Ruolin. I'm aware of Unity's coordinate system convention, which is why I believe that the Y-axis should be modified in the unit tests (instead of the Z-axis) when testing `TurnLeft` and `TurnRight`. (Note that in my original message the sign might be wrong, though.) Is there some conversion going on under the hood and the python code that we work with does not use Unity's convention?

As for the Euler angles, when I look at
```python
def euler_angles_allclose(euler1, euler2, atol = 1e-6):
    """Compare two euler angles."""
    wrapped_angles1 = [wrap_angle(degrees_to_radians(a)) for a in euler1]
    wrapped_angles2 = [wrap_angle(degrees_to_radians(a)) for a in euler2]
    dists = [get_signed_angle_distance(a, b)
             for a, b in zip(wrapped_angles1, wrapped_angles2)]
    return np.allclose(dists, 0, atol=atol)
```
, the input values `euler1` and `euler2` are converted to *radians*, then their difference is computed, and then `np.allclose` is used but with `atol` in *degrees* (like when it is used in the left turn scenario: `assert euler_angles_allclose(new_robot_base_rotation, expected_robot_base_rotation, atol=5.0`). Am I misunderstanding it?

**Reply by thisjustin123** [reply]
Hi @martinschonger, I'm a bit new to this new RCareWorld, but I'll still try to help as much as I can. I'll try to make a new branch to tweak these tests' correctness, since you bring up some good points that, unless I'm also missing something, show our tests are not actually verifying correct behavior.

From what I can tell...

> Specifically, the left turn does not work at all here. Why is that?

You are saying here that you are observing the right turn cause the Y-angle to rotate, is that correct? I see that in your print output the Y-axis is the only one significantly changing after you rotated 20 degrees right. It does look to me like we should be testing the Y axis. It does make sense anyways, given that Unity has a Y-up-coordinate system. 

> The input values euler1 and euler2 are converted to radians, then their difference is computed.

I do recognize the discrepancy you're seeing here... it seems like we're testing if the radians error tolerance falls within 5 RADIANS, which is practically 360 degrees, rendering this test useless. This is definitely a symptom of our lack of documentation.

I'll try to test this out myself and tweak some things. Will continue on [this branch](https://github.com/empriselab/RCareWorld/tree/justin/test-stretch-bathing-fix) and update on progress.

**Reply by thisjustin123** [reply]
I just made a PR #122 to fix at least the tests. It seems as though, as discussed, the robot movement is in fact not lining up with even the fixed test cases, since the branch's PR unit tests are still failing.

**Reply by thisjustin123** [reply]
Thinking about it more, I think this bug is happening due to variable performance of everyone's machines. I'm fairly certain it's caused by the fact that Unity may run slower/quicker depending on framerate, meaning the simulation can literally get slowed down vs. real time with lower framerates.

I'm not exactly sure why this is happening but I'll look into a fix.

**Reply by DanielTakeshi** [reply]
@thisjustin123 Thanks for the comment. Do you have any empirical results or minimum working example code (ideally run on separate machines) to demonstrate?

**Reply by martinschonger** [reply]
Hi Justin, thanks for looking into this.

> You are saying here that you are observing the right turn cause the Y-angle to rotate, is that correct?

Yes. I can see you fixed that in the unit tests.

> I think this bug is happening due to variable performance of everyone's machines. I'm fairly certain it's caused by the fact that Unity may run slower/quicker depending on framerate, meaning the simulation can literally get slowed down vs. real time with lower framerates.

I tested on a more powerful machine and there the left and right turn behavior is closer to what I would expect. This is interesting since at least the physics simulation should be framerate-independent according to https://docs.unity3d.com/Manual/ExecutionOrder.html. Still, I guess this discrepancy means that a controller designed on some machine would behave differently on the evaluation platform, right?

**Reply by DanielTakeshi** [reply]
@martinschonger Thanks for checking. But indeed the last point you bring up is critical. How may we tell if our solution(s) are going to work well if it could be dependent on something like the machine it's evaluated on? We would need to know exactly the machine specs, and if we don't have that machine, presumably we're out of luck?

**Reply by thisjustin123** [reply]
I should be able to change the logic in Unity to make it framerate independent. Currently, it runs off the logic that the sim *should* be framerate independent, which for some reason doesn't seem to be the case despite our use of `FixedUpdate`, and then just runs it for a fixed time duration. We can change it to instead try to meet some new rotation and just keep rotating until it hits that value, which should truly be framerate independent.

Just wanted to give a heads up that we'll look into this soon! We'll try to fix this up in the next week.

**Comment by thisjustin123** [comment]
Just a heads up--I am working on this now. Here's a demo of the rotation being a lot more precise than before:

(Roughly 90-degree left turn)

https://github.com/user-attachments/assets/133d2e26-2b94-461a-bd9e-08076b95b89b

Still testing and tuning. Thanks for your patience!

**Reply by ALaws0n** [reply]
Hello, 

In the video above, are you completing the turn by only calling `TurnLeft(90, speed)` on the robot, or are you also moving the position of the gripper?

**Reply by YoruCathy** [reply, marked as answer]
Hi @ALaws0n , we've updated the simulation environment and the example in the latest commit. Check it out!
